examples/coappearances.cpp

```diff
@@ -0,0 +1,327 @@
+#include "picojson.h"
+
+#include <flatdata/flatdata.h>
+#include <fstream>
+#include <iostream>
+#include <string>
+
+namespace co = coappearances;
+
+void
+convert_meta( const picojson::object& obj, co::GraphBuilder builder, std::string& strings )
+{
+    // Since flatdata's mutators are not holding any data, we are creating a vector with a single
+    // element for holding the data.
+    flatdata::Vector< co::Meta > data( 1 );
+    auto meta = data[ 0 ];
+    meta.title_ref = strings.size( );
+    strings += obj.at( "title" ).get< std::string >( ) + '\0';
+    meta.author_ref = strings.size( );
+    strings += obj.at( "author" ).get< std::string >( ) + '\0';
+    builder.set_meta( meta );
+}
+
+uint16_t
+convert_id( const std::string& id )
+{
+    assert( id.size( ) == 2 );
+    return ( static_cast< uint8_t >( id[ 0 ] ) << 8 ) + static_cast< uint8_t >( id[ 1 ] );
+}
+
+using CharactersIndex = std::map< uint16_t /* id */, uint16_t /* ref */ >;
+
+CharactersIndex
+build_characters_index( const picojson::object& characters )
+{
+    CharactersIndex index;
+    if ( std::numeric_limits< uint16_t >::max( ) <= characters.size( ) )
+    {
+        throw std::runtime_error( "Too many characters for indexing by 16 bits" );
+    }
+
+    uint16_t ref = 0;
+    for ( const auto& kv : characters )
+    {
+        uint16_t id = convert_id( kv.first );
+        if ( index.count( id ) )
+        {
+            throw std::runtime_error( std::string( "Duplicate index " ) + kv.first );
+        }
+        index.emplace( id, ref );
+        ref++;
+    }
+    return index;
+}
+
+std::pair< uint8_t /* major */, uint8_t /* minor */ >
+convert_chapter( const std::string& s )
+{
+    auto pos = s.find( '.' );
+    if ( pos == std::string::npos )
+    {
+        throw std::runtime_error( std::string( "Unexpected chapter format for " ) + s );
+    }
+
+    auto major = std::stoul( s.substr( 0, pos ) );
+    if ( ( 1 << co::Chapter::MajorType::bit_width ) <= major )
+    {
+        throw std::runtime_error( "Major chapter overflow" );
+    }
+
+    auto minor = std::stoul( s.substr( pos + 1 ) );
+    if ( ( 1 << co::Chapter::MinorType::bit_width ) <= minor )
+    {
+        throw std::runtime_error( "Major chapter overflow" );
+    }
+
+    return {static_cast< uint8_t >( major ), static_cast< uint8_t >( minor )};
+}
+
+void
+convert_characters( const picojson::object& characters,
+                    const CharactersIndex& characters_index,
+                    co::GraphBuilder builder,
+                    std::string& strings )
+{
+    auto vertices = builder.start_vertices( );
+    auto vertices_data = builder.start_vertices_data( );
+
+    for ( const auto& kv : characters )
+    {
+        const auto& id = kv.first;
+        const auto& data = kv.second.get< picojson::object >( );
+
+        auto character = vertices.grow( );
+        character.name_ref = strings.size( );
+        strings += data.at( "name" ).get< std::string >( ) + '\0';
```

Comment:
This is pretty straight-forward, but still too error-prone. It is very easy to forget to add `+ '\0'` and then the reader would read complete garbage (happened to me several times). What do you think about adding a simple helper class? The interfaces could be similar to `flatdata::Vector`: `grow` and `close`.